### PR TITLE
Feat/chat service

### PR DESCRIPTION
**Pull request by kaladivo**

### CODE REVIEW

File: apps/metrics-service/src/index.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import {runMainInNode} from '@vexl-next/server-utils/src/runMainInNode'
 import {metricsServer} from './metricsServer'
 
+console.log('pid: ', process.pid)
```
Comment (kaladivo) [inline apps/metrics-service/src/index.ts]: remove

File: apps/chat-service/src/db/ChallegeDbService/queries/createUpdateChallengeInvalidate.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export const createUpdateChallengeInvalidate = Effect.gen(function* (_) {
+  const sql = yield* _(PgClient.PgClient)
+
+  const resolver = yield* _(
+    SqlResolver.void('updateChallengeInvalidate', {
+      Request: ChatChallengeE,
+      execute: (params) => sql`
+        UPDATE challenge
+        SET
+          valid = FALSE
```
Comment (SamTremko) [inline apps/chat-service/src/db/ChallegeDbService/queries/createUpdateChallengeInvalidate.ts:16]: delete challenge if possible

File: apps/chat-service/src/routes/inbox/createInbox.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      yield* _(validateChallengeInBody(req.body))
+
+      const inboxService = yield* _(InboxDbService)
+      const hashedPublicKey = yield* _(hashPublicKey(req.body.publicKey))
+
+      const existingInbox = yield* _(
+        inboxService.findInboxByPublicKey(hashedPublicKey)
+      )
+      if (Option.isSome(existingInbox)) {
+        return {}
```
Comment (SamTremko) [inline apps/chat-service/src/routes/inbox/createInbox.ts]: if existing update client version and platform

File: apps/chat-service/src/routes/inbox/deleteInboxes.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    Effect.all(
+      Array.map(req.body.dataForRemoval, (inboxToDelete) =>
+        Effect.gen(function* (_) {
+          yield* _(validateChallengeInBody(inboxToDelete))
+          const hashedPublicKey = yield* _(
+            hashPublicKey(inboxToDelete.publicKey)
+          )
+
+          const inboxService = yield* _(InboxDbService)
+          // TODO delete all messages - can be done in non blocking fashion
```
Comment (SamTremko) [inline apps/chat-service/src/routes/inbox/deleteInboxes.ts]: make sure to do this

File: apps/chat-service/src/routes/inbox/requestApproval.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    if (Option.isNone(whitelistRecordOption)) {
+      return true
+    }
+
+    const whitelistRecord = whitelistRecordOption.value
+
+    if (
+      whitelistRecord.state === 'CANCELED' ||
+      whitelistRecord.state === 'WAITING'
```
Comment (SamTremko) [inline apps/chat-service/src/routes/inbox/requestApproval.ts]: what if LEFT

File: apps/chat-service/src/routes/inbox/requestApproval.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    makeEndpointEffect(
+      Effect.gen(function* (_) {
+        const {receiverInbox, senderInbox} = yield* _(
+          findAndEnsureReceiverAndSenderInbox({
+            receiver: req.body.publicKey,
+            sender: sec['public-key'],
+          })
+        )
+
+        if (!canSendSendRequest({receiverInbox, senderInbox})) {
```
Comment (SamTremko) [inline apps/chat-service/src/routes/inbox/requestApproval.ts]: typo

File: apps/chat-service/src/routes/inbox/updateInbox.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import {UpdateInboxEndpoint} from '@vexl-next/rest-api/src/services/chat/specification'
+import {Effect} from 'effect'
+import {Handler} from 'effect-http'
+
+// Depreciated - left here for backwards compatibility
```
Comment (SamTremko) [inline apps/chat-service/src/routes/inbox/updateInbox.ts:5]: Update client version

File: packages/rest-api/src/services/chat/specification.ts [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  'createInbox',
+  '/api/v1/inboxes'
+).pipe(
+  Api.setSecurity(ServerSecurity),
+  Api.setResponseStatus(200 as const),
+  Api.setRequestBody(CreateInboxRequestE),
+  Api.setRequestHeaders(CommonHeaders),
+  Api.setResponseBody(CreateInboxResponseE)
+)
+
```
Comment (SamTremko) [inline packages/rest-api/src/services/chat/specification.ts:67]: pls move to contracts

**Review by SamTremko — APPROVED** [review]

File: apps/chat-service/src/routes/inbox/requestApproval.ts (hunk 5 above)
Comment (kaladivo) [reply apps/chat-service/src/routes/inbox/requestApproval.ts]: In the end there is no left state (in that case the whitelist record is not present). But I forgot about DISAPROVED state. That should also be there.

File: apps/chat-service/src/routes/inbox/updateInbox.ts (hunk 7 above)
Comment (kaladivo) [reply apps/chat-service/src/routes/inbox/updateInbox.ts:5]: Decided to put this into the retrieveMessages endpoint.

File: apps/chat-service/src/__tests__/routes/approveRequest.test.ts [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+                    publicKeyToConfirm: generatePrivateKey().publicKeyPemBase64,
+                    approve: true,
+                  })
+                ),
+              },
+              HttpClientRequest.setHeaders(user2.authHeaders)
+            ),
+            Effect.either
+          )
+          expectErrorResponse(SenderInboxDoesNotExistError)(errorResponse)
```
Comment (SamTremko) [inline apps/chat-service/src/__tests__/routes/approveRequest.test.ts:273]: pls make the name more clear

File: apps/chat-service/src/__tests__/routes/cancelRequest.test.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+            ),
+            Effect.either
+          )
+
+          expectErrorResponse(RequestNotPendingError)(failedReqResponse)
+        })
+      )
+    })
+
+    it('receiver inbox does not exist', async () => {
```
Comment (SamTremko) [inline apps/chat-service/src/__tests__/routes/cancelRequest.test.ts]: make sure ist not duplicate

File: apps/chat-service/src/__tests__/routes/deleteInboxes.test.ts [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+                dataForRemoval: [
+                  yield* _(user2.inbox1.addChallenge({})),
+                  yield* _(user2.inbox2.addChallenge({})),
+                ],
+              },
+            },
+            HttpClientRequest.setHeaders(user2.authHeaders)
+          )
+        )
+        expectInboxDeletedFully(String(id1))
```
Comment (SamTremko) [inline apps/chat-service/src/__tests__/routes/deleteInboxes.test.ts]: yield pls

**Review by SamTremko — APPROVED** [review]

**Pull request merged**